KKUMI-42 apple login

eekrwl opened the pull request:
## 구현내용
### 1. 애플 로그인
애플 로그인을 구현했습니다. 그동안 해봤던 구글, 카카오 로그인과 조금 달랐습니다.
<img width="772" alt="스크린샷 2024-07-22 오전 2 25 03" src="https://github.com/user-attachments/assets/d616d36b-d208-4185-9b7a-236ba00c164f">
1. 클라이언트로부터 authorization code 받아오기
2. client secret을 서버에서 생성
3. 생성한 client secret과 여러 정보들로 유저 정보 요청

client secret을 직접 생성해서 쓴다는 게 특이한 점이었던 것 같습니다.

## 고민사항
### 1. SigninRequest를 쪼개야할까?
카카오 로그인에서는 클라이언트로부터 accessToken, refreshToken을 받아오는데,
애플 로그인에서는 authorization code를 받게되어 기존의 token들만 있던 dto를 변경해야할지 고민했습니다.
굳이 파일을 더 늘릴 필요가 있을까 싶어 아래와 같이 authorizationCode만 추가해주었습니다.
```java
public class SigninRequest {

    private String refreshToken;
    private String accessToken;
    private String authorizationCode;
}
```

AuthService 코드에서 아래와 같이 분리해서 토큰이라는 이름으로 가져왔습니다.
```java
//필요한 토큰 가져오기
String token = null;
if (oAuthProvider == OAuthProvider.KAKAO) { //카카오는 엑세스 토큰
    token = request.getAccessToken();
} else if (oAuthProvider == OAuthProvider.APPLE) { //애플은 인가코드
    token = request.getAuthorizationCode();
}
```

### 2. 같은 이메일 같은 유저?
여러 OAuth를 통해 가져온 유저의 이메일이 같다면 같은 기존 유저로 볼지 고민입니다.
현재 구현된 모습을 생각해보면, 이메일로 유저가 우리서비스에 가입되어 있는지 판단하고 있습니다. 그 결과, 애플로 로그인하여 사용하던 유저가 처음으로 카카오로 로그인을 시도했을 때, 이메일이 같아서 그냥 기존 같은 유저로 인식합니다. 한개의 이메일을 여러 사람이 사용할 수는 없을 것 같아 괜찮을 것 같기도합니다. 하지만, 서비스 사용자로서 여러 서비스를 사용했던 걸 생각해보면 **네이버로 로그인되어 있습니다. 카카오로도 연결하시겠습니까?** 같은 알림이 떴던 것 같아 고민중입니다. 이메일이 같으면 어차피 같은 유저인데 왜 한번더 연결하겠냐고 확인하는지 찾아봐야 할 것 같습니다.

eekrwl committed:
KKUMI-42 [FEAT] #27 애플 로그인을 위한 properties 추가

AWS secrets manager에 데이터를 추가하고 연결해주었습니다.

eekrwl committed:
KKUMI-42 [CHORE] #27 bouncycastle, jackson 등 의존성 추가

eekrwl committed:
KKUMI-42 [FEAT] #27 애플에서 인가코드로 유저정보 받아오기

client secret을 직접 생성해줬습니다.

eekrwl committed:
KKUMI-42 [FEAT] #27 애플 로그인 기능

애플 로그인 기능을 추가했습니다.
카카오 로그인에는 accessToken, refreshToken이 필요하고, 애플 로그인에는 authorization code가 필요해서 SigninRequest의 모양을 바꿔야하나 고민하던 중, 그냥 authorizationCode 항목을 추가해주었습니다. 필요한 정보만 넣어서 전달하면되고, 파일이 더 쪼개지지 않아 보기 편했습니다.

babjo commented:
> 1. SigninRequest를 쪼개야할까?
카카오 로그인에서는 클라이언트로부터 accessToken, refreshToken을 받아오는데,
애플 로그인에서는 authorization code를 받게되어 기존의 token들만 있던 dto를 변경해야할지 고민했습니다.
굳이 파일을 더 늘릴 필요가 있을까 싶어 아래와 같이 authorizationCode만 추가해주었습니다.

넵 쪼개는게 다른 사람이 코드를 읽을 때 혼동이 없습니다. 아니면 같이 파일에 쓰고 javadoc 남기는 것을 추천합니다.

> 2. 같은 이메일 같은 유저?
여러 OAuth를 통해 가져온 유저의 이메일이 같다면 같은 기존 유저로 볼지 고민입니다.
현재 구현된 모습을 생각해보면, 이메일로 유저가 우리서비스에 가입되어 있는지 판단하고 있습니다. 그 결과, 애플로 로그인하여 사용하던 유저가 처음으로 카카오로 로그인을 시도했을 때, 이메일이 같아서 그냥 기존 같은 유저로 인식합니다. 한개의 이메일을 여러 사람이 사용할 수는 없을 것 같아 괜찮을 것 같기도합니다. 하지만, 서비스 사용자로서 여러 서비스를 사용했던 걸 생각해보면 네이버로 로그인되어 있습니다. 카카오로도 연결하시겠습니까? 같은 알림이 떴던 것 같아 고민중입니다. 이메일이 같으면 어차피 같은 유저인데 왜 한번더 연결하겠냐고 확인하는지 찾아봐야 할 것 같습니다.

넵넵

eekrwl commented:
> > 1. SigninRequest를 쪼개야할까?
> >    카카오 로그인에서는 클라이언트로부터 accessToken, refreshToken을 받아오는데,
> >    애플 로그인에서는 authorization code를 받게되어 기존의 token들만 있던 dto를 변경해야할지 고민했습니다.
> >    굳이 파일을 더 늘릴 필요가 있을까 싶어 아래와 같이 authorizationCode만 추가해주었습니다.
> 
> 넵 쪼개는게 다른 사람이 코드를 읽을 때 혼동이 없습니다. 아니면 같이 파일에 쓰고 javadoc 남기는 것을 추천합니다.

앗 혼란을 줄 여지가 있군요 수정하겠습니다!

eekrwl committed:
KKUMI-42 [REFACTOR] #27 SigninRequest -> Apple, Kakao로 분리

하나의 SigninRequest에 refreshToken, accessToken, authorization code를 다 넣었었는데, 보는 사람에게 혼란을 줄 수 있다고 하여 수정했습니다.

eekrwl committed:
KKUMI-42 [FIX] #27 User테이블 Provider에 무조건 KAKAO로 들어가던 문제 해결
